rust/worker/src/execution/operators/metadata_filtering.rs

```diff
@@ -35,75 +55,353 @@ impl MetadataFilteringOperator {
         query_ids: Option<Vec<String>>,
+        where_clause: Option<Where>,
+        offset: Option<u32>,
+        limit: Option<u32>,
     ) -> Self {
         Self {
-            log_record,
+            blockfile_provider,
             record_segment,
             metadata_segment,
-            blockfile_provider,
-            where_clause,
-            where_document_clause,
+            log_record,
             query_ids,
+            where_clause,
+            offset,
+            limit,
         }
     }
 }
 
 #[derive(Debug)]
 pub(crate) struct MetadataFilteringOutput {
     pub(crate) log_records: Chunk<LogRecord>,
-    // Offset Ids of documents that match the where and where_document clauses.
-    pub(crate) where_condition_filtered_offset_ids: Option<Vec<u32>>,
-    // Offset ids of documents that the user specified in the query directly.
-    pub(crate) user_supplied_filtered_offset_ids: Option<Vec<u32>>,
+    pub(crate) log_mask: RoaringBitmap,
+    pub(crate) offset_ids: RoaringBitmap,
 }
 
 #[derive(Error, Debug)]
 pub(crate) enum MetadataFilteringError {
     #[error("Error creating record segment reader {0}")]
-    MetadataFilteringRecordSegmentReaderCreationError(#[from] RecordSegmentReaderCreationError),
+    RecordSegmentReaderCreationError(#[from] RecordSegmentReaderCreationError),
     #[error("Error materializing logs {0}")]
-    MetadataFilteringLogMaterializationError(#[from] LogMaterializerError),
+    LogMaterializationError(#[from] LogMaterializerError),
     #[error("Error filtering documents by where or where_document clauses {0}")]
-    MetadataFilteringIndexError(#[from] MetadataIndexError),
+    IndexError(#[from] MetadataIndexError),
     #[error("Error from metadata segment reader {0}")]
-    MetadataFilteringMetadataSegmentReaderError(#[from] MetadataSegmentError),
+    MetadataSegmentReaderError(#[from] MetadataSegmentError),
     #[error("Error reading from record segment")]
-    MetadataFilteringRecordSegmentReaderError,
+    RecordSegmentReaderError,
     #[error("Invalid input")]
-    MetadataFilteringInvalidInput,
+    InvalidInput,
 }
 
 impl ChromaError for MetadataFilteringError {
     fn code(&self) -> ErrorCodes {
         match self {
-            MetadataFilteringError::MetadataFilteringRecordSegmentReaderCreationError(e) => {
-                e.code()
+            MetadataFilteringError::RecordSegmentReaderCreationError(e) => e.code(),
+            MetadataFilteringError::LogMaterializationError(e) => e.code(),
+            MetadataFilteringError::IndexError(e) => e.code(),
+            MetadataFilteringError::MetadataSegmentReaderError(e) => e.code(),
+            MetadataFilteringError::RecordSegmentReaderError => ErrorCodes::Internal,
+            MetadataFilteringError::InvalidInput => ErrorCodes::InvalidArgument,
+        }
+    }
+}
+
+pub(crate) struct MetadataLogReader<'me> {
+    compact_metadata: BTreeMap<&'me str, BTreeMap<&'me MetadataValue, RoaringBitmap>>,
```

Comment:
same as below - the outer map can be a hashmap instead of a btreemap

Reply:
Even the inner map does not have to be strictly a btreemap as we don't have a use-case for dynamically inserting in a sorted fashion. We could keep a `Vec<(&MetadataValue, RoaringBitmap)>`, push all the data to it, sort it and keep it here. Then on demand, we could search using binary search

Reply:
Just to add more color to not prefering a btreemap here - btreemaps generally have overhead for e.g. they replicate the key 6 ways. No strong opinion here though